x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/PersistentCacheTests.java

```diff
@@ -172,4 +193,214 @@ public void testCleanUp() throws Exception {
+                        final long sizeInCacheFile = randomCacheFile.getCompletedRanges()
+                            .stream()
+                            .mapToLong(range -> range.v2() - range.v1())
+                            .sum();
+                        assertThat(
+                            "Persistent cache should contain cached data for at least 1 cache file",
+                            sizeInCache,
+                            greaterThanOrEqualTo(sizeInCacheFile)
+                        );
+
+                        final CacheFile.EvictionListener listener = evictedCacheFile -> {};
+                        randomCacheFile.acquire(listener);
+                        try {
+                            SortedSet<Tuple<Long, Long>> ranges = null;
+                            while (ranges == null || ranges.isEmpty()) {
+                                ranges = randomPopulateAndReads(randomCacheFile);
+                            }
+                            assertTrue(cacheService.isCacheFileToSync(randomCacheFile));
+                        } finally {
+                            randomCacheFile.release(listener);
+                        }
+
+                        assertThat(
+                            "Persistent cache should contain cached data for at least 1 cache file",
+                            persistentCache.getCacheSize(cacheKey.getShardId(), snapshotId),
+                            equalTo(sizeInCache)
+                        );
+                    }
+                }
+
+                final boolean fsyncFailure = true;
+                randomBoolean();
+                logger.debug("blocking fsync for cache file [{}] with failure [{}]", randomCacheFile.getFile(), fsyncFailure);
+                fileSystem.blockFSyncForPath(randomCacheFile.getFile(), fsyncFailure);
+
+                logger.debug("starting synchronization of cache files");
+                final Thread fsyncThread = new Thread(cacheService::synchronizeCache);
+                fsyncThread.start();
+
+                logger.debug("waiting for synchronization of cache files to be blocked");
+                fileSystem.waitForBlock();
+
+                logger.debug("starting eviction of shard [{}]", cacheKey);
+                cacheService.markShardAsEvictedInCache(cacheKey.getSnapshotUUID(), cacheKey.getSnapshotIndexName(), cacheKey.getShardId());
+
+                logger.debug("waiting for shard eviction to be processed");
+                cacheService.waitForCacheFilesEvictionIfNeeded(
+                    cacheKey.getSnapshotUUID(),
+                    cacheKey.getSnapshotIndexName(),
+                    cacheKey.getShardId()
+                );
+
+                logger.debug("unblocking synchronization of cache files");
+                fileSystem.unblock();
+                fsyncThread.join();
+
+                assertThat(
```

Comment:
Can we also verify that num-docs decreased, just to verify that some docs were actually deleted from the index? Since we now have the file-exists check, the check in this line could pass just because the files are gone.

Reply:
The number of docs returned by `PersistentCache#getNumDocs()` concerns the number of Lucene documents which can also contain deleted documents. I pushed 7e88cd244db00464677f51701a24fb77a3134092 to compute the cache size without taking into account the file-exists check.